### DISCUSSION TITLE
Use 'bun:sqlite' as main driver in 'Bun.build' bundler

### DISCUSSION THREAD

**Discussion by Borderliner** [mikro-orm/mikro-orm#5700] [General]
Hi, I'd like to know if it's possible to use bun's native sqlite driver, instead of sqlite3 or better-sqlite3. It's giving me these errors when I'm trying to use `build.ts`:
```bash
$ bun build --compile --target=bun ./src/index.ts --outfile myapp 
7 |     return require('better-sqlite3');
                       ^
error: Could not resolve: "better-sqlite3". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\better-sqlite3\index.js:7:20

63 |     return require('pg');
                        ^
error: Could not resolve: "pg". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\postgres\index.js:63:20

192 |       : require('pg-query-stream');
                      ^
error: Could not resolve: "pg-query-stream". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\postgres\index.js:192:17

94 |     const tds = require('tedious');
                             ^
error: Could not resolve: "tedious". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\mssql\index.js:94:25

24 |     return require('mysql');
                        ^
error: Could not resolve: "mysql". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\mysql\index.js:24:20

10 |         return require('mariadb/callback');
                            ^
error: Could not resolve: "mariadb/callback". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\@mikro-orm\knex\dialects\mysql\MariaDbKnexDialect.js:10:24

8 |         return require('better-sqlite3');
                           ^
error: Could not resolve: "better-sqlite3". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\@mikro-orm\knex\dialects\sqlite\BetterSqliteKnexDialect.js:8:24

11 |         return require('libsql');
                            ^
error: Could not resolve: "libsql". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\@mikro-orm\knex\dialects\sqlite\LibSqlKnexDialect.js:11:24

44 |     return require('pg');
                        ^
error: Could not resolve: "pg". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\redshift\index.js:44:20

13 |     return require('pg').native;
                        ^
error: Could not resolve: "pg". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\pgnative\index.js:13:20

44 |     const oracledb = require('oracledb');
                                  ^
error: Could not resolve: "oracledb". Maybe you need to "bun install"?
    at C:\Users\X\Projects\smm_est\server\node_modules\knex\lib\dialects\oracledb\index.js:44:30
```

Build file:
```typescript
await Bun.build({
    entrypoints: ["./src/index.ts"],
    outdir: "./build",
    plugins: [],
    target: 'bun',
    minify: true
})
```

**Comment by B4nan** [comment]
You would have to create your own driver, which should be super simple for sqlite since you can reuse pretty much all the code (the main different would be in the knex dialect which does the require call https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/dialects/sqlite/LibSqlKnexDialect.ts). 

But that is a different story, your problem is that bun is trying to bundle all the optional peer dependencies of knex, you will need to configure it somehow to skip them. Usually there is a way to specify external dependencies to get around this.

**Reply by B4nan** [reply]
Here is an issue on their end:

https://github.com/oven-sh/bun/issues/7886

AnThis one is also pretty much the same but about different dependency. Maybe you could try this workaround:

https://github.com/oven-sh/bun/issues/4803#issuecomment-2020683735

**Reply by B4nan** [reply]
cc @co-sic not sure if you cracked this one, you are using some SQL driver with bun, right?

**Reply by co-sic** [reply]
I am running a bun service with mikro-orm and the postgresql driver, but i never used the Bundler from bun. Since it is used in a backend service, i'm just running the ts files (as recommended by them) in a Docker container. Issue seems to be related only to bundling.

**Reply by Borderliner** [reply]
@B4nan Thanks, I'll try the fix and report.
@co-sic Yes, it's only an issue with the bundler. The app itself runs fine. I was just experimenting. It's nice to have the bundler work as well.

**Reply by vitalyso** [reply]
@B4nan thanks! Had a problem with bundling with MikroORM + Postgres driver with bun, the workaround you shared did help!

**Reply by ceopaludetto** [reply]
> You would have to create your own driver, which should be super simple for sqlite since you can reuse pretty much all the code (the main different would be in the knex dialect which does the require call https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/dialects/sqlite/LibSqlKnexDialect.ts).
> 
> But that is a different story, your problem is that bun is trying to bundle all the optional peer dependencies of knex, you will need to configure it somehow to skip them. Usually there is a way to specify external dependencies to get around this.

If someone wants to use `bun:sqlite` before official support, I've managed to make it work by patching `@mikro-orm/libsql`. To do the same, just follow the steps below:

1. Run `bun patch @mikro-orm/knex`
2. Open `node_modules/@mikro-orm/knex/dialects/sqlite/LibSqlKnexDialect.js`
3. Do the following modifications:

```patch
diff --git a/dialects/sqlite/LibSqlKnexDialect.js b/dialects/sqlite/LibSqlKnexDialect.js
index 21a2d277aa6b7ffce5901af09a17fe837596c2d5..c833abf6e2fccb79ca591b7e97064d1641e1ec16 100644
--- a/dialects/sqlite/LibSqlKnexDialect.js
+++ b/dialects/sqlite/LibSqlKnexDialect.js
@@ -5,10 +5,10 @@ const SqliteTableCompiler_1 = require("./SqliteTableCompiler");
 const MonkeyPatchable_1 = require("../../MonkeyPatchable");
 class LibSqlKnexDialect extends MonkeyPatchable_1.MonkeyPatchable.BetterSqlite3Dialect {
     get driverName() {
-        return 'libsql';
+        return 'bun:sqlite';
     }
     _driver() {
-        return require('libsql');
+        return require('bun:sqlite').Database;
     }
     async _query(connection, obj) {
         /* istanbul ignore next */
@@ -24,16 +24,10 @@ class LibSqlKnexDialect extends MonkeyPatchable_1.MonkeyPatchable.BetterSqlite3D
         const bindings = this._formatBindings(obj.bindings);
         const response = await statement[callMethod](bindings);
         obj.response = response;
-        obj.context = {
-            lastID: response.lastInsertRowid,
-            changes: response.changes,
-        };
         return obj;
     }
     async acquireRawConnection() {
-        const connection = new this.driver(this.connectionSettings.filename, {
-            ...this.connectionSettings,
-        });
+        const connection = new this.driver(this.connectionSettings.filename);
         connection.__created = Date.now();
         return connection;
     }
```

4. Run `bun patch --commit 'node_modules/@mikro-orm/knex'`
5. Enjoy!

---

> [!WARNING]  
> This doesn't ensure that everything will work (I don't know if one of the libraries have some sort of edge case that breaks the other one)

> [!NOTE]
> I'll try to develop a officially supported bun package later, but I don't know how to write tests that run in bun context using Jest in the current monorepo configuration (this problem breaks a lot of tests locally)

**Comment by mikelgmh** [comment]
I'm facing the same issue when trying to build my app. Running the app in dev mode works fine.
